w3lib/url.py

```diff
@@ -393,95 +393,98 @@ def _safe_ParseResult(parts, encoding='utf8', path_encoding='utf8'):
-    'http://www.example.com/r%C3%A9sum%C3%A9'
-    >>>
-
-    For more examples, see the tests in `tests/test_url.py`.
-    """
-    # If supplied `encoding` is not compatible with all characters in `url`,
-    # fallback to UTF-8 as safety net.
-    # UTF-8 can handle all Unicode characters,
-    # so we should be covered regarding URL normalization,
-    # if not for proper URL expected by remote website.
-    try:
-        scheme, netloc, path, params, query, fragment = _safe_ParseResult(
-            parse_url(url), encoding=encoding)
-    except UnicodeEncodeError as e:
-        scheme, netloc, path, params, query, fragment = _safe_ParseResult(
-            parse_url(url), encoding='utf8')
-
-    # 1. decode query-string as UTF-8 (or keep raw bytes),
-    #    sort values,
-    #    and percent-encode them back
-    if six.PY2:
-        keyvals = parse_qsl(query, keep_blank_values)
-    else:
-        # Python3's urllib.parse.parse_qsl does not work as wanted
-        # for percent-encoded characters that do not match passed encoding,
-        # they get lost.
-        #
-        # e.g., 'q=b%a3' becomes [('q', 'b\ufffd')]
-        # (ie. with 'REPLACEMENT CHARACTER' (U+FFFD),
-        #      instead of \xa3 that you get with Python2's parse_qsl)
-        #
-        # what we want here is to keep raw bytes, and percent encode them
-        # so as to preserve whatever encoding what originally used.
-        #
-        # See https://tools.ietf.org/html/rfc3987#section-6.4:
-        #
-        # For example, it is possible to have a URI reference of
-        # "http://www.example.org/r%E9sum%E9.xml#r%C3%A9sum%C3%A9", where the
-        # document name is encoded in iso-8859-1 based on server settings, but
-        # where the fragment identifier is encoded in UTF-8 according to
-        # [XPointer]. The IRI corresponding to the above URI would be (in XML
-        # notation)
-        # "http://www.example.org/r%E9sum%E9.xml#r&#xE9;sum&#xE9;".
-        # Similar considerations apply to query parts.  The functionality of
-        # IRIs (namely, to be able to include non-ASCII characters) can only be
-        # used if the query part is encoded in UTF-8.
-        keyvals = parse_qsl_to_bytes(query, keep_blank_values)
-    keyvals.sort()
-    query = urlencode(keyvals)
-
-    # 2. decode percent-encoded sequences in path as UTF-8 (or keep raw bytes)
-    #    and percent-encode path again (this normalizes to upper-case %XX)
-    uqp = _unquotepath(path)
-    path = quote(uqp, _safe_chars) or '/'
-
-    fragment = '' if not keep_fragments else fragment
-
-    # every part should be safe already
-    return urlunparse((scheme,
-                       netloc.lower().rstrip(':'),
-                       path,
-                       params,
-                       query,
-                       fragment))
+try:
+    from scurl import canonicalize_url
+except ImportError as e:
```

Comment:
if we run `python -c "from scurl import canonicalize_url"` in tox then probably it will print out `ImportError` for those envs that don't have scurl installed right?

Reply:
>  it will print out ImportError for those envs that don't have scurl installed right?

Yes, I meant to do this only when running scurl tests, so that we know that scurl at least imports, and it's *supposed* to be used (would be great to know that for sure though).

Reply:
what if we use `echo` to print out the message we need in tox, @lopuhin ? 😄

Reply:
@nctl144 I see what you mean by `echo` now, this is not what I had in mind :)
Right now we have this bit in ``w3lib.url``:
```
try:
    from scurl import canonicalize_url
except ImportError as e:
    canonicalize_url = _canonicalize_url
```
Now suppose that in ``scurl`` environment this import ``from scurl import canonicalize_url`` fails for some reason: e.g. the scurl library installed but failed to import due to some ``.so`` issues, or we moved the ``canonicalize_url``, or some other reason. But we won't notice this in the tests as they stand now, because w3lib will fall back to the python implementation, and all tests will pass. So I would like for tests in ``scurl`` environment to fail if we fall back to pure python w3lib. One way to do it is to add ``python -c "from scurl import canonicalize_url"`` to the scurl environments. Or maybe even better would be to make sure that ``w3lib.url.canonicalize_url`` is coming from ``scurl``, but will probably need a separate small script.

Reply:
oh I see your point now @lopuhin , I will just add the command `python -c "from scurl import canonicalize_url"` for now since it's the shortest way to detect import failure. But we will see what people's opinions on this are 😄